### PR TITLE
add capture and replay feature in tensorrt

### PR DESCRIPTION
**Pull request by lanluo-nvidia**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 1]
```diff
@@ -32,6 +34,7 @@ def __init__(
         capture_fx_graph_before: Optional[List[str]] = None,
         capture_fx_graph_after: Optional[List[str]] = None,
         save_engine_profile: bool = False,
+        capture_shim: bool = False,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py]: Rename this something like `capture_tensorrt_api_recording`.

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 2]
```diff
@@ -65,6 +70,7 @@ def __init__(
         self.cfg = DebuggerConfig(
             log_level=log_level,
             save_engine_profile=save_engine_profile,
+            capture_shim=capture_shim,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py]: Same here

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 3]
```diff
@@ -92,6 +98,26 @@ def __init__(
         self.capture_fx_graph_before = capture_fx_graph_before
         self.capture_fx_graph_after = capture_fx_graph_after
 
+        if self.cfg.capture_shim:
+            if not sys.platform.startswith("linux"):
+                _LOGGER.warning(
+                    "capture_shim featureis only supported on linux, will disable it"
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py]: `Capturing TensorRT API calls is only supported on Linux, therefore ignoring the capture_tensorrt_api_recording setting`

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 4]
```diff
@@ -92,6 +98,26 @@ def __init__(
...
+                )
+                self.cfg.capture_shim = False
+                return
+            if ENABLED_FEATURES.tensorrt_rtx:
+                _LOGGER.warning(
+                    "capture_shim feature is not supported on TensorRT-RTX, will disable it"
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py]: Capturing TensorRT API calls is only supported with standard TensorRT, therefore ignoring the capture_tensorrt_api_recording setting

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 5]
```diff
@@ -92,6 +98,26 @@ def __init__(
...
+                )
+                self.cfg.capture_shim = False
+                return
+            if not is_tensorrt_version_supported("10.13.0"):
+                _LOGGER.warning(
+                    "capture_shim feature is only supported on TensorRT 10.13 and above, will disable it"
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py]: Capturing TensorRT API calls is only supported for TensorRT versions 10.13 and above, therefore ignoring the capture_tensorrt_api_recording setting

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 6]
```diff
@@ -143,6 +169,46 @@ def __enter__(self) -> None:
...
+                    _LOGGER.error(
+                        f"LD_PRELOAD is not set, please add the {shim_lib_name} with full path to the LD_PRELOAD environment variable"
+                    )
+                    is_valid = False
+                if os.environ.get("TRT_SHIM_NVINFER_LIB_NAME") is None:
+                    _LOGGER.error(
+                        f"TRT_SHIM_NVINFER_LIB_NAME is not set, please add the {nvinfer_lib_name} with full path to the TRT_SHIM_NVINFER_LIB_NAME environment variable"
+                    )
+                    is_valid = False
+                if os.environ.get("TRT_SHIM_OUTPUT_JSON_FILE") is None:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py]: This should just go into the debugging dir

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 7]
```diff
@@ -143,6 +169,46 @@ def __enter__(self) -> None:
...
+                    _LOGGER.error(
+                        "TRT_SHIM_OUTPUT_JSON_FILE is not set, please add the shim output json file name with full path to the TRT_SHIM_OUTPUT_JSON_FILE environment variable"
+                    )
+                    is_valid = False
+                else:
+                    shim_output_json_file = os.environ["TRT_SHIM_OUTPUT_JSON_FILE"]
+                    shim_output_dir = os.path.dirname(shim_output_json_file)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py]: Same thing here, we have a debugging directory that gets created for this stuff

File: tools/debug/capture_replay/README.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
+## Introduction
```
Comment (narendasan) [inline tools/debug/capture_replay/README.md]: Can we add this to the web documentation?

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 9]
```diff
@@ -91,6 +96,10 @@ def __init__(
 
         self.capture_fx_graph_before = capture_fx_graph_before
         self.capture_fx_graph_after = capture_fx_graph_after
+        if os.environ.get("TORCHTRT_ENABLE_TENSORRT_API_CAPTURE") == "1":
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py]: This will get weird, not sure what the right way to go is. 

I think the correct behavior would be 

#### Case 1

```py
with torch_tensorrt.Debugger(capture_tensorrt_api_recording=True):
```

but the script was invoked with 

```sh
python my_script.py
```

Should result in a warning like `WARNING: In order to capture TensorRT API calls, please invoke the script with environment variable TORCHTRT_ENABLE_TENSORRT_API_CAPTURE=1` 

And if the user does this: 

#### Case 2

```py
with torch_tensorrt.Debugger(capture_tensorrt_api_recording=False):
```

but the script was invoked with 

```sh
TORCHTRT_ENABLE_TENSORRT_API_CAPTURE=1 python my_script.py
```

Then we do nothing. Don't override their setting just a log INFO message in `torch_tensorrt.__init__` that we loaded the shim library. 

And if the user does this 

#### Case 3 

```py
with torch_tensorrt.Debugger(capture_tensorrt_api_recording=1):
```

but the script was envoked with 

```sh
TORCHTRT_ENABLE_TENSORRT_API_CAPTURE=1 python my_script.py
```

Then it should run normally

File: py/torch_tensorrt/__init__.py [hunk 10]
```diff
@@ -24,6 +26,69 @@
 
 import torch
 
+
```
Comment (narendasan) [inline py/torch_tensorrt/__init__.py]: Do you think this should be with the other TRT proxy code?

Comment (narendasan) [reply py/torch_tensorrt/__init__.py]: That way we can directly control the import order

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 11]
```diff
@@ -91,6 +96,10 @@ def __init__(
...
+            self.cfg.capture_tensorrt_api_recording = True
+        else:
+            self.cfg.capture_tensorrt_api_recording = False
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py:117]: We still need the behavior to move the default temp location for the location to the debugging folder

File: py/torch_tensorrt/dynamo/debug/_Debugger.py [hunk 12]
```diff
@@ -92,6 +98,18 @@ def __init__(
         self.capture_fx_graph_before = capture_fx_graph_before
         self.capture_fx_graph_after = capture_fx_graph_after
 
+        if self.cfg.capture_tensorrt_api_recording:
+            env_flag = os.environ.get("TORCHTRT_ENABLE_TENSORRT_API_CAPTURE", None)
+            if env_flag is None or (env_flag != "1" and env_flag.lower() != "true"):
+                # currently this feature is only supported for TensorRT on Linux platform
+                if (
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/debug/_Debugger.py]: I think we can make this more meaningful to the user, like errors for the unsupported conditions and the warning for the rest. 

So like:

```py
if self.cfg.capture_tensorrt_api_recording:
    if ENABLED_FEATURES.tensorrt_rtx:
        _LOGGER.error("TensorRT API capture is not supported with TensorRT-RTX"
    if not sys.platform.startswith("linux"):
        _LOGGER.error("TensorRT API capture is only supported on Linux")
    if env_flag is None or (env_flag != "1" and env_flag.lower() != "true"):
        _LOGGER.warning(
                        "In order to capture TensorRT API calls, please invoke the script with environment variable TORCHTRT_ENABLE_TENSORRT_API_CAPTURE=1"
                    )
    else:
        print("...find the logs here...")
```

File: py/torch_tensorrt/_TensorRTProxyModule.py [hunk 13]
```diff
@@ -28,6 +32,66 @@ def _find_lib(name: str, paths: List[str]) -> str:
...
+                ctypes.CDLL(
+                    os.path.join(path, "libtensorrt_shim.so"), mode=ctypes.RTLD_GLOBAL
+                )
+                tensorrt_lib_path = path
+                break
+            except Exception as e:
+                continue
+
+    if tensorrt_lib_path is None:
+        _LOGGER.error(
```
Comment (narendasan) [inline py/torch_tensorrt/_TensorRTProxyModule.py]: Maybe this should be a warning?

File: tools/debug/capture_replay/README.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+Run the engine with `trtexec`:
+
+```bash
+trtexec --loadEngine=/absolute/path/to/output_engine
+```
+
+### Notes
+- Ensure the `libnvinfer.so` used by the shim matches the TensorRT version in your environment.
+- If multiple TensorRT versions are installed, prefer absolute paths as shown above.
+- The capture is best-effort; if your program builds multiple engines, multiple captures may be produced.
```
Comment (narendasan) [inline tools/debug/capture_replay/README.md]: This isnt the behavior right? Also make this file into an rst and move it into docsrc
